File: api/applib/src/main/java/org/apache/causeway/applib/layout/grid/Grid.java
/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package org.apache.causeway.applib.layout.grid;

import java.io.Serializable;
import java.util.Map;
import java.util.stream.Stream;

import org.apache.causeway.applib.annotation.Programmatic;
import org.apache.causeway.applib.layout.component.ActionLayoutData;
import org.apache.causeway.applib.layout.component.CollectionLayoutData;
import org.apache.causeway.applib.layout.component.PropertyLayoutData;
import org.apache.causeway.applib.services.layout.LayoutService;

/**
 * All top-level page layout classes should implement this interface.
 *
 * <p>It is used by the {@link LayoutService} as a common based type for any layouts read in from XML.
 *
 * @since 1.x revised for 4.0 {@index}
 */
@Deprecated
@Programmatic
public interface Grid {

    Class<?> domainClass();

    /**
     * Arbitrary additional 'runtime' data attributed to this grid,
     * but not part of the DTO specification.
     * @since 4.0
     */
    Map<String, Serializable> attributes();

    /**
     * Indicates whether or not this grid is a fallback.
     * {@code True}, if this Grid originates from
     * {@link org.apache.causeway.applib.services.grid.GridSystemService#defaultGrid(Class)}.
     * <p>
     * Governs meta-model facet precedence, that is,
     * facets from annotations should overrule those from fallback XML grids.
     */
    boolean isFallback();
    boolean isNormalized();

    Stream<PropertyLayoutData> streamPropertyLayoutData();
    Stream<CollectionLayoutData> streamCollectionLayoutData();
    Stream<ActionLayoutData> streamActionLayoutData();

}

File: incubator/clients/kroviz/src/main/kotlin/org/apache/isis/client/kroviz/to/bs3/Grid.kt
/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package org.apache.isis.client.kroviz.to.bs3

import org.apache.isis.client.kroviz.to.TransferObject
import org.w3c.dom.Document
import org.w3c.dom.Node
import org.w3c.dom.asList

/**
 * For the Wicket Viewer the following layout is used:
 * * rows[0] (head) contains the object title and actions
 * * rows[1] contains data, tabs, collections, etc.
 * * there may be N other rows as well
 * Please note, that rows may be children of Tab as well (recursive)
 */
class Grid(document: Document) : TransferObject {
    var rows = ArrayList<Row>()

    init {
        val root = document.firstChild!!
        val kids = root.childNodes
        val rowNodes = kids.asList()
        val rowList = rowNodes.filter { it.nodeName.equals("bs3:row") }
        for (n: Node in rowList) {
            val row = Row(n)
            rows.add(row)
        }
    }

}
